nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+    in_progress: BTreeMap<SagaId, slog::Logger>,
+    succeeded: Vec<RecoverySuccess>,
+    failed: Vec<RecoveryFailure>,
+}
+
+impl ExecutionBuilder {
+    pub fn new() -> ExecutionBuilder {
+        ExecutionBuilder {
+            in_progress: BTreeMap::new(),
+            succeeded: Vec::new(),
+            failed: Vec::new(),
+        }
+    }
+
+    pub fn build(self) -> Execution {
+        assert!(
+            self.in_progress.is_empty(),
+            "attempted to build execution result while some recoveries are \
+            still in progress"
+        );
+        Execution { succeeded: self.succeeded, failed: self.failed }
+    }
+
+    /// Record that we've started recovering this saga
+    pub fn saga_recovery_start(
+        &mut self,
+        saga_id: SagaId,
+        saga_logger: slog::Logger,
+    ) {
+        info!(&saga_logger, "recovering saga: start");
+        assert!(self.in_progress.insert(saga_id, saga_logger).is_none());
+    }
+
+    /// Record that we've successfully recovered this saga
+    pub fn saga_recovery_success(&mut self, saga_id: SagaId) {
+        let saga_logger = self
+            .in_progress
+            .remove(&saga_id)
+            .expect("recovered saga should have previously started");
+        info!(saga_logger, "recovered saga");
+        self.succeeded.push(RecoverySuccess { time: Utc::now(), saga_id });
+    }
+
+    /// Record that we failed to recover this saga
+    pub fn saga_recovery_failure(&mut self, saga_id: SagaId, error: &Error) {
+        let saga_logger = self
+            .in_progress
+            .remove(&saga_id)
+            .expect("recovered saga should have previously started");
+        error!(saga_logger, "failed to recover saga"; error);
+        self.failed.push(RecoveryFailure {
+            time: Utc::now(),
+            saga_id,
+            message: InlineErrorChain::new(error).to_string(),
+        });
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::status;
+    use crate::test::make_fake_saga;
+    use crate::test::make_saga_ids;
+    use omicron_test_utils::dev::test_setup_log;
+
+    #[test]
+    fn test_read_all_from_channel() {
+        let (tx, mut rx) = mpsc::unbounded_channel();
+
+        // If we send nothing on the channel, reading from it should return
+        // immediately, having found nothing.
+        let (numbers, disconnected) = read_all_from_channel::<u32>(&mut rx);
+        assert!(numbers.is_empty());
+        assert!(!disconnected);
+
+        // Send some numbers and make sure we get them back.
+        let expected_numbers = vec![1, 7, 0, 1];
+        for n in &expected_numbers {
+            tx.send(*n).unwrap();
+        }
+        let (numbers, disconnected) = read_all_from_channel(&mut rx);
+        assert_eq!(expected_numbers, numbers);
+        assert!(!disconnected);
+
+        // Send some more numbers and make sure we get them back.
+        let expected_numbers = vec![9, 7, 2, 0, 0, 6];
```

Comment:
this kinda feels like it's within sigma of a property test...not that the behavior is so complex or input-dependent that it _needs_ to be, though...